Kill transposes and tensordots, part 1

amilsted opened the pull request:
This attempts to avoid unnecessary ops by:

1. Only calling `tensordot()` once per `contract_between()`.
2. Replacing `tensordot()` with a version that minimizes transposes.

Although `tensordot()` is not horribly complicated, adding this is not insignificant. Ultimately, perhaps the changes to tensordot should head upstream to TF? Also, I don't know if including this code requires any additional license notices or somesuch.

Would appreciate a review on this one!

Thenerdstation commented on tensornetwork.py:
```diff
@@ -521,60 +523,59 @@ def _remove_trace_edge(self, edge: Edge, new_node: Node) -> None:
+  
+    remaining_edges = []
+    for (i, edge) in enumerate(node1_edges):
+      if edge not in edges:  # FIXME: Makes the cost quadratic in # edges
```
cast edges to a set() before this loop and use that.

amilsted replied:
Done.

Thenerdstation commented on tensornetwork.py:
```diff
@@ -521,60 +523,59 @@ def _remove_trace_edge(self, edge: Edge, new_node: Node) -> None:
 
-  def _remove_edge(self, edge: Edge, new_node: Node) -> None:
-    """Collapse an edge in the network.
+  def _remove_edges(self, edges, node1: Node, node2: Node, new_node: Node) -> None:
```
Add typing

amilsted replied:
Done.

Thenerdstation commented on tensornetwork.py:
```diff
@@ -521,60 +523,59 @@ def _remove_trace_edge(self, edge: Edge, new_node: Node) -> None:
-  def _remove_edge(self, edge: Edge, new_node: Node) -> None:
-    """Collapse an edge in the network.
+  def _remove_edges(self, edges, node1: Node, node2: Node, new_node: Node) -> None:
+    """Collapse a list of edges shared by two nodes in the network.
```
Passing the edges and the nodes is redundant. Just pass the edges

amilsted replied:
Actually, it's so that the ordering of the arguments to tensordot can be specified. Previously, the ordering was `edge.node1, edge.node2`. Now we have multiple edges, so there's no unique choice based on them. Instead, we use the arguments to e.g. `contract_between` to decide the ordering. This data must be passed on to `reorder_edges`, which needs to know which axes of `new_node` correspond to which axes of the old nodes.

Also, one may wish to choose the ordering to minimize the need for transposes of the output tensor.

Thenerdstation commented on tensornetwork.py:
```diff
@@ -902,14 +917,38 @@ def contract_between(self,
+    axes2 = []
+    for edge in shared_edges:
+      if edge.node1 is node1:
+        axes1.append(edge.axis1)
```
Add a comment explaining why this if else is needed. (I know why, but it might not be obvious for new people).

Thenerdstation commented on tensordot2.py:
```diff
@@ -0,0 +1,213 @@
+from __future__ import print_function
+
+import numpy as np
+from six.moves import xrange  # pylint: disable=redefined-builtin
```
Remove this (we plan on only supporting python3).

amilsted replied:
Ok. This was leftover from the TF copy-paste :)

Thenerdstation commented on tensordot2.py:
```diff
@@ -0,0 +1,213 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# FIXME: This is a modified version of code from the TensorFlow project.
```
Let's avoid any legal commentary in code.

amilsted replied:
Killed it.

Thenerdstation commented on tensordot2.py:
```diff
@@ -0,0 +1,213 @@
+
+import tensorflow as tf
+
+def tensordot(a, b, axes, name=None):
```
This is really good! Can you include some benchmarks of both graph building and runtime of using this?

amilsted replied:
I have some benchmarks now, but am not sure where to put them. Are you suggesting they should be listed in a docstring or comment?

amilsted commented:
Should I add separate tests for tensordot2? I could put in a (subset of) the tests used for tf.tensordot(), which I have been using myself to check I don't break stuff.

Thenerdstation commented:
Yes, please add unit tests for tensordot2.

Thenerdstation commented:
I'm curious how this performs when compared to `einsum`.

amilsted commented:
Rebased on current master. Added tests for tensordot2. There is a problem with the tests: pytest doesn't like `test_tensordot_scalar_axes()` for some reason I don't understand. @Thenerdstation?

Thenerdstation commented:
Yeah we'll likely have to modify the tests heavily. You're also importing modules directly from TF when there are public accessors to those methods.

Would you be ok with my me pushing commits to your branch? I think I could fix this fairly quickly.

amilsted commented:
Sure, go ahead!

Thenerdstation commented:
Yeah you were working way too hard. pytest is amazing, so always assume it can do what you need.

Thenerdstation commented:
Ignore that, realized the tests weren't running. Still gotta fix some stuff...

Thenerdstation commented:
Got it working. Tests are much cleaner now. 

We're gonna wait on #52 before we merge this. You'll likely have to fix some merge conflicts.

amilsted commented:
Thanks. I dressed up my benchmarks for tensordot2 a little. Here they are: https://gist.github.com/amilsted/dea5a8ad4b0b928657e96052ef13a9ed

Up to around 20% performance improvement in run time and a consistently large (~40%) improvement in build time. All test cases are instances where tensordot() adds (sometimes trivial) transpose ops, but tensordot2 never adds a transpose.

Thenerdstation commented:
Awesome! Good enough evidence for me.

On Fri, May 17, 2019, 1:16 PM Ashley Milsted <notifications@github.com>
wrote:

> Thanks. I dressed up my benchmarks for tensordot2 a little. Here they are:
> https://gist.github.com/amilsted/dea5a8ad4b0b928657e96052ef13a9ed
>
> Up to around 20% performance improvement in run time and a consistently
> large (~40%) improvement in build time. All test cases are instances where
> tensordot() adds (sometimes trivial) transpose ops, but tensordot2 never
> adds a transpose.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/TensorNetwork/pull/6?email_source=notifications&email_token=ACCMQWRVD7GVAUOM65VMMTDPV4HDJA5CNFSM4HJQZPEKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVVYLEA#issuecomment-493585808>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACCMQWSGQHFSZNAOOMLHVODPV4HDJANCNFSM4HJQZPEA>
> .
>

Thenerdstation commented:
I confirmed on my own colab that adding a tensordot op went from take 10.7 ms to 6.99ms, so build times are 1.5x faster. 

This is awesome! I'm so happy right now.

amilsted commented:
By the way, although this PR removes the call to `flatten_edges()` in `contract_between()`, we might want to consider some of the tricks that are in the *original* tensordot() (as well as tensordot2) for minimizing the shape-related graph ops it produces, in case someone uses it by hand. By this I mean things like avoiding the use of `tf.shape()` if the shape is fully defined at build time.

amilsted commented:
> Yeah you were working way too hard. pytest is amazing, so always assume it can do what you need.

Hey, don't look at me :) This was mostly copy-pasted from the TF source!

Thenerdstation commented:
You'll have to integrate this with our new `TensorFlowBackend`. It should be pretty straight forward, but if you need help feel free to shoot me an email.

amilsted commented:
Tests pass, but pylint is being a pain. Do all of the pylint messages need addressing? At least one of them is a "len-as-condition" whose "fixing" breaks the tests (sometimes the sequences is a numpy array). Some messages are from untouched TF parts of tensordot(). How should I go about this?

amilsted commented:
> I'm curious how this performs when compared to `einsum`.

Judging by the TF source, einsum is quite good at avoiding unnecessary ops (it is better at avoiding unneeded reshapes than tensordot(2)), but is not smart enough to use the transpose flags of matmul to avoid tf.tranpose() ops, so it still ought to work out slower in the benchmark cases I tested.

Thenerdstation commented:
> Tests pass, but pylint is being a pain.

Just add `# pylint: disable=...`. we've already had to do that a few times.

Thenerdstation commented:
Did some quick fixes to the pylint. Hope you don't mind.
